enhancements/machine-config/manage-boot-images.md

```diff
@@ -282,44 +285,108 @@ spec:
 }
 
-// MachineSetBootImageHistoryStatus defines the observed state of MachineSetBootImageHistory
-type MachineSetBootImageHistoryStatus struct {
+type MachineResourceReference struct {
+	// name is the machine management resource's name
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+	// kind is the machine management resource's kind
+	// +kubebuilder:validation:Required
+	// +required
+	Kind string `json:"kind"`
+	// apiGroup is name of the APIGroup that the machine management resource belongs to. This is for disambiguating
+	// between Cluster API and Machine API backed resources.
+	// +kubebuilder:validation:Required
+	// +required
+	APIGroup string `json:"apiGroup"`
 }
 
 // BootImageHistoryDetail is the struct for each element in the Details array
 type BootImageHistoryDetail struct {
-	Index          int         `json:"index"`
-	UpdatedTime    metav1.Time `json:"updatedTime"`
-	BootImageRef   string      `json:"bootImageRef"`
+	// updateTime records the timestamp at which the update took place.
+	// +required
+	UpdateTime metav1.Time `json:"updatedTime"`
+	// bootImageRef records the new boot image reference to which the update took place.
+	// +required
+	BootImageRef string `json:"bootImageRef"`
 }
 
-// MachineSetBootImageHistoryList contains a list of MachineSetBootImageHistory
-type MachineSetBootImageHistoryList struct {
+// BootImageHistoryList contains a list of BootImageHistory
+type BootImageHistoryList struct {
 	metav1.TypeMeta `json:",inline"`
 	metav1.ListMeta `json:"metadata,omitempty"`
-	Items           []MachineSetBootImageHistory `json:"items"`
+	Items           []BootImageHistory `json:"items"`
 }
+
 ```
-There will be one instance of this per `Machineset`. It will be updated by the MSBIC as `Machinesets` are created/updated and will exist in the same namespace as the `MachineSet`. This CRD will also need to support MAPI and CAPI backed `MachineSets`. The goal of this is to provide information about the "lineage" of a `MachineSet` to the user. The user can then manually restore their `MachineSet` to an earlier state if they wish to do so by following documentation. 
+There will be one instance of this per machine management resource(which can be a MachineSet[MAPI or CAPI], MachineDeployment...etc). It will be named the same as the resource being tracked. The MSBIC is responsible for creating and updating this CR when a boot image update takes place. This CR will exist in the same namespace as the resource.
+
+YAML Example for a MAPI backed machineset scenario:
+```
+apiVersion: machineconfiguration.openshift.io/v1alpha1
+kind: BootImageHistory
+metadata:
+  name: djoshy10-2tcqv-worker-a
+spec: {}
+status:
+  machineResourceReference:
+    name: djoshy10-2tcqv-worker-a
+    kind: MachineSet
+    apiGroup: cluster.x-k8s.io/v1alpha3
+  details:
+    - updateTime: "2023-12-14T12:00:00Z"
+      bootImageRef: "projects/rhcos-cloud/global/images/rhcos-414-92-202308032115-0-gcp-x86-64"
```

Comment:
Is the boot image ref always going to be a single string?

Reply:
The source of this data conceptually is coreos "stream metadata" which are uniquely identified by version strings (that also get embedded in some cases in the infrastructure-specified fields like this one for GCP).

Today as folks know this is stored in a configmap updated by the installer.

However, one thing we could do is have $something (maybe the installer again) make cluster-local versioned snapshots of the configmap on in-place update.

Then `BootImageHistory` could just reference that API object.

To answer the concrete question:

> Is the boot image ref always going to be a single string?

In most cases that's "good enough" and where it's not in theory we could squash to a single string, but I think my preference would be having this data always canonically stored in "stream metadata" format which is at least a well-defined and understood thing - and other things are just pointers into it.  (Well except for the machineset API objects because they aren't coreos specific they already end up with this data in another format; but we can't change that)

Reply:
For now, I'm going to opt to do both the [configmap generation](https://github.com/openshift/enhancements/pull/1496#discussion_r1457726874) and [the RHCOS version string](https://github.com/openshift/enhancements/pull/1496#discussion_r1449328333) approaches mentioned in the below threads. I think there is value in both. As the expectation here is that the customer will do some manual work to restore their old boot images(via documentation and the backup data from this CR), I don't think it's a step too far if the customer has to do some math to figure out the original "stream metadata" object that their cluster installer had originally pointed to.